File: Level 3/[카카오 인턴] 경주로 건설/solution.py
from collections import deque


def is_valid(x, y, N):
    return 0 <= x < N and 0 <= y < N


def is_blocked(x, y, N, board):
    return (x, y) == (0, 0) or not is_valid(x, y, N) or board[x][y] == 1


def calculate_cost(direction, prev_direction, cost):
    if prev_direction == -1 or prev_direction == direction:
        return cost + 100
    else:
        return cost + 600


def is_should_update(x, y, direction, new_cost, visited):
    return visited[x][y][direction] == 0 or visited[x][y][direction] > new_cost


def solution(board):
    N = len(board)
    directions = [(0, -1), (-1, 0), (0, 1), (1, 0)]
    visited = [[[0 for _ in range(len(directions))] for _ in range(N)] for _ in range(N)]
    queue = deque([(0, 0, -1, 0)])

    answer = float("inf")

    while queue:
        x, y, prev_direction, cost = queue.popleft()

        for direction, (dx, dy) in enumerate(directions):
            nx, ny = x + dx, y + dy

            if is_blocked(nx, ny, N, board):
                continue

            new_cost = calculate_cost(direction, prev_direction, cost)

            if (nx, ny) == (N - 1, N - 1):
                answer = min(answer, new_cost)
            elif is_should_update(nx, ny, direction, new_cost, visited):
                queue.append((nx, ny, direction, new_cost))
                visited[nx][ny][direction] = new_cost

    return answer

File: Level 3/[카카오 인턴] 경주로 건설/solution.js
function solution(board) {
  const N = board.length;
  const directions = [
    [0, -1],
    [-1, 0],
    [0, 1],
    [1, 0],
  ];
  const visited = Array.from({ length: N }, () =>
    Array.from({ length: N }, () => Array(directions.length).fill(0))
  );
  const queue = [[0, 0, -1, 0]];

  let answer = Infinity;

  while (queue.length) {
    const [x, y, prevDirection, cost] = queue.shift();

    directions.forEach(([dx, dy], direction) => {
      const [nx, ny] = [x + dx, y + dy];

      if (isBlocked(nx, ny, N, board)) {
        return;
      }

      const newCost = calculateCost(direction, prevDirection, cost);

      if (nx === N - 1 && ny === N - 1) {
        answer = Math.min(answer, newCost);
      } else if (isShouldUpdate(x, y, direction, newCost, visited)) {
        visited[x][y][direction] = newCost;
        queue.push([nx, ny, direction, newCost]);
      }
    });
  }

  return answer;
}

function isValid(x, y, N) {
  return x >= 0 && x < N && y >= 0 && y < N;
}

function isBlocked(x, y, N, board) {
  return (x === 0 && y === 0) || !isValid(x, y, N) || board[x][y] === 1;
}

function calculateCost(direction, prevDirection, cost) {
  return prevDirection === -1 || prevDirection === direction
    ? cost + 100
    : cost + 600;
}

function isShouldUpdate(x, y, direction, newCost, visited) {
  return visited[x][y][direction] === 0 || visited[x][y][direction] > newCost;
}
